libs/unity/library/Runtime/Scripts/Media/VideoRenderer.cs

```diff
@@ -15,23 +16,27 @@ namespace Microsoft.MixedReality.WebRTC.Unity
     /// via the attached <a href="https://docs.unity3d.com/ScriptReference/Renderer.html">Renderer</a>.
     /// </remarks>
     [RequireComponent(typeof(Renderer))]
-    [AddComponentMenu("MixedReality-WebRTC/Media Player")]
-    public class MediaPlayer : MonoBehaviour
+    [AddComponentMenu("MixedReality-WebRTC/Video Renderer")]
+    public class VideoRenderer : MonoBehaviour
     {
-        [Header("Source")]
-        [Tooltip("Audio source providing the audio data used by this player")]
-        [SerializeField]
-        protected MediaSource AudioSource;
+        // Note - Here what we really want is to serialize some reference to an IVideoSource.
+        // Unfortunately Unity before 2019.3 does not support serialized interfaces, and more generally ignores
+        // polymorphism in serialization. And although there are ways to make that work with a custom inspector,
+        // this disables support for object picker, so only drag-and-drop works, which is very impractical.
+        // So we use a class deriving from MonoBehaviour, which is the only entity for which Unity handles polymorphism.
+        // This makes the UI slightly confusing because we cannot filter audio vs. video sources, since C# supports only
+        // single inheritance and we already need to divide senders from receivers, so cannot also split by media kind.
+        // Instead we use OnValidate() to make sure the media source is a video source.
```

Comment:
There are a bit too many classes IMHO and this seems to be one of the causes.

Can we split the hierarchy of track sources and have:
```
MonoBehaviour
|-VideoProducer
  |-VideoTrackSource
  |-VideoReceiver
|-AudioTrackSource
| AudioReceiver
```
? This would take out `MediaTrackSource` (which is almost empty) and `MediaReceiver` (which could be replaced by an interface maybe?) AFAICS these are only used by MediaLine, which could use MonoBehavior for serialized fields instead and cast to interfaces to call methods... But this doesn't sound too great. 

Any better ideas?

Reply:
You're basically throwing away any type safety for the entire transceiver API.

Reply:
How exactly? Isn't the UI for MediaLine gating assignments of sources/receivers? I admit serializing a MonoBehavior looks ugly but AFAICS the potential for introducing errors is limited (I might be wrong, haven't looked at all the code yet).

My main issue is that splitting logic across a long hierarchy of tiny classes with similar names is confusing since responsibilities are not clearly defined. If we need `MediaProducer` and `MediaTrackSource` for the sake of type-checking these serialized fields, maybe we can make this explicit and remove all logic from them, making them explicit placeholders? Specific suggestions on the class contents:
- the `MediaKind` field seems unused, and if used it could simply be an abstract property rather than a field with a chain of constructors initializing it
- the logic in `MediaProducer` handles a work queue, has nothing to do with media; maybe rename to WorkQueue and have `MediaProducer (empty) : WorkQueue : MonoBehaviour`?
- the logic in `MediaTrackSource` is superfluous - it would be simpler to define OnEnable()/OnDisable() in the subclasses directly and keep this class empty.

Reply:
> You're basically throwing away any type safety for the entire transceiver API.

Also note that type safety is already far from watertight - you can pass an Audio producer to VideoRenderer and an AudioTrackSource to a video MediaLine and vice versa 😛